flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/DataStreamJoin.scala

```diff
@@ -0,0 +1,180 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.nodes.datastream
+
+import org.apache.calcite.plan._
+import org.apache.calcite.rel.core.JoinRelType
+import org.apache.calcite.rel.{BiRel, RelNode, RelWriter}
+import org.apache.flink.api.common.functions.RichFilterFunction
+import org.apache.flink.api.java.functions.NullByteKeySelector
+import org.apache.flink.streaming.api.datastream.DataStream
+import org.apache.flink.table.api.{StreamTableEnvironment, TableException}
+import org.apache.flink.table.calcite.FlinkTypeFactory
+import org.apache.flink.table.plan.nodes.CommonJoin
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalJoin
+import org.apache.flink.table.plan.schema.RowSchema
+import org.apache.flink.table.runtime.join.{JoinUtil, ProcTimeInnerJoin}
+import org.apache.flink.table.runtime.types.{CRow, CRowTypeInfo}
+import org.apache.flink.types.Row
+
+/**
+  * Flink RelNode which matches along with JoinOperator and its related operations.
+  */
+class DataStreamJoin(
+   cluster: RelOptCluster,
+   traitSet: RelTraitSet,
+   leftNode: RelNode,
+   rightNode: RelNode,
+   joinNode: FlinkLogicalJoin,
+   leftSchema: RowSchema,
+   schema: RowSchema,
+   ruleDescription: String)
+  extends BiRel(cluster, traitSet, leftNode, rightNode)
+  with CommonJoin
+  with DataStreamRel {
+
+  override def deriveRowType() = schema.logicalType
+
+  override def copy(traitSet: RelTraitSet, inputs: java.util.List[RelNode]): RelNode = {
+    new DataStreamJoin(
+      cluster,
+      traitSet,
+      inputs.get(0),
+      inputs.get(1),
+      joinNode,
+      leftSchema,
+      schema,
+      ruleDescription)
+  }
+
+  override def toString: String = {
+
+    s"${joinTypeToString(joinNode.getJoinType)}" +
+      s"(condition: (${joinConditionToString(schema.logicalType,
+        joinNode.getCondition, getExpressionString)}), " +
+      s"select: (${joinSelectionToString(schema.logicalType)}))"
+  }
+
+  override def explainTerms(pw: RelWriter): RelWriter = {
+    super.explainTerms(pw)
+      .item("condition", joinConditionToString(schema.logicalType,
+        joinNode.getCondition, getExpressionString))
+      .item("select", joinSelectionToString(schema.logicalType))
+      .item("joinType", joinTypeToString(joinNode.getJoinType))
+  }
+
+  override def translateToPlan(tableEnv: StreamTableEnvironment): DataStream[CRow] = {
+
+    val config = tableEnv.getConfig
+
+    // get the equality keys and other condition
+    val (leftKeys, rightKeys, otherCondition) =
+      JoinUtil.analyzeJoinCondition(joinNode, getExpressionString)
+
+    if (left.isInstanceOf[StreamTableSourceScan]
```

Comment:
@fhueske @hongyuhong 
I believe the logic is that we need to differentiate in this backbone class between the cases of JOIN: Stream - Stream and Stream - Table, and then have a condition for each. I believe this case is to restrict (for now) the implementation only for the Stream-Stream.
The question is now - will it always be the case that if we create an SQL query, that has one of the inputs a table - this will be of "StreamTableSourceScan" ...is it possible to have it of a different type?